### PR TITLE
8295964: Move InstanceKlass::_misc_flags

### PR DESCRIPTION
**Pull request by coleenp**
I moved misc_flags out to their own misc flags class so that we can put the writeable accessFlags there too with atomic access.

Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295964](https://bugs.openjdk.org/browse/JDK-8295964): Move InstanceKlass::_misc_flags

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10249/head:pull/10249` \
`$ git checkout pull/10249`

Update a local copy of the PR: \
`$ git checkout pull/10249` \
`$ git pull https://git.openjdk.org/jdk pull/10249/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10249`

View PR using the GUI difftool: \
`$ git pr show -t 10249`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10249.diff">https://git.openjdk.org/jdk/pull/10249.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp [hunk 1]
```diff
@@ -4403,7 +4403,9 @@ void VM_RedefineClasses::redefine_single_class(Thread* current, jclass the_jclas
     scratch_class->enclosing_method_method_index());
   scratch_class->set_enclosing_method_indices(old_class_idx, old_method_idx);
 
-  the_class->set_has_been_redefined();
+  if (!the_class->has_been_redefined()) {
+    the_class->set_has_been_redefined();
+  }
```
Comment (sspitsyn) [inline src/hotspot/share/prims/jvmtiRedefineClasses.cpp:4408]: Is this change really needed?

**Review by sspitsyn — APPROVED** [review]
Nice refactoring and simplification.
It looks good to me.
Thanks,
Serguei

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/InstanceKlass.java [hunk 2]
```diff
@@ -124,18 +111,6 @@ private static synchronized void initialize(TypeDataBase db) throws WrongTypeExc
     CLASS_STATE_BEING_INITIALIZED = db.lookupIntConstant("InstanceKlass::being_initialized").intValue();
     CLASS_STATE_FULLY_INITIALIZED = db.lookupIntConstant("InstanceKlass::fully_initialized").intValue();
     CLASS_STATE_INITIALIZATION_ERROR = db.lookupIntConstant("InstanceKlass::initialization_error").intValue();
-
-    MISC_REWRITTEN                    = db.lookupIntConstant("InstanceKlass::_misc_rewritten").intValue();
-    MISC_HAS_NONSTATIC_FIELDS         = db.lookupIntConstant("InstanceKlass::_misc_has_nonstatic_fields").intValue();
```
Comment (dholmes-ora) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/InstanceKlass.java:129]: These all seem to have been deleted rather than modified ??

**Review by dholmes-ora — APPROVED** [review]
Refactoring looks good. These are all the immutable flags right? (ie set once during class creation, and no concurrency issues).

One query on SA below.

Thanks.

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/prims/jvmtiRedefineClasses.cpp:4408]: This is needed by this change.  The has_been_redefined flag can be set at runtime and with these flags, there's an assert for all that they're only set once.  This one didn't have the assert but it can only be set in a safepoint, so I have this code to make it an exception. For now.  The plan is to make the _flags set once, and the _status field (to be added) set at runtime, and this can be moved to that.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/InstanceKlass.java (hunk 2 above)
Comment (coleenp) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/InstanceKlass.java:129]: Yes, they are unused by the SA and not really needed for core dump analysis unless someone finds a reason to see these.  If so, some support can be added at that time.  A better approach would be to have some python code to decode this in gdb if needed (probably only for scratch_class).

**Comment by coleenp** [comment]
Thank you for the code reviews, David and Serguei.
